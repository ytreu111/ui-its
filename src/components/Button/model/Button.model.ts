import { ReactNode } from "react";
import CSS from "csstype";

import IPadding from "types/interface/IPadding";
import IMargin from "types/interface/IMargin";

export interface IButtonProps extends IPadding, IMargin{
  type?: ButtonType;
  size?: SizeType;
  disabled?: boolean;
  buttonStyle?: StyleType;
  icon?: ReactNode;
  iconSize?: IIconSize | number;
  iconColor?: CSS.Property.Color;
  iconMargin?: number;
  loading?: boolean;
  onClick?: (e: MouseEvent) => void;
  clickEffect?: boolean;
  rounded?: boolean;
  children?: ReactNode;
  fontSize?: CSS.Property.FontSize;
  width?: CSS.Property.Width;
}

interface IIconSize {
  width: CSS.Property.Width;
  height: CSS.Property.Height;
}

export enum ButtonTypeEnum {
  primary = 'primary',
  danger = 'danger',
  orange = 'orange',
  yellow = 'yellow',
  green = 'green',
  purple = 'purple',
  magenta = 'magenta',
  blue = 'blue',
  transparent = 'transparent',
}

export enum ButtonSizeEnum {
  large = 'large',
  default = 'default',
  small = 'small',
}

export enum ButtonStyleEnum {
  //TODO Временно убрано по просьбе дизайнеров, тк оно не используется в системе
  // stroke = 'stroke',
  default = 'default',
  translucent = 'translucent',
  transparent = 'transparent',
}

export type ButtonType = keyof typeof ButtonTypeEnum;

export type SizeType = keyof typeof ButtonSizeEnum;

export type StyleType = keyof typeof ButtonStyleEnum;