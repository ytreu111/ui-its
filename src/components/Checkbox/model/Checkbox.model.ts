import { ChangeEvent, ReactElement } from "react";

export interface ICheckboxProps {
  checked?: boolean;
  value?: any;
  moved?: boolean;
  disabled?: boolean;
  checkboxSize?: number;
  name?: string;
  iconRender?: (checked: boolean) => ReactElement;
  onChange?: (checked: boolean, value: any, event: ChangeEvent<HTMLInputElement>) => void;
}