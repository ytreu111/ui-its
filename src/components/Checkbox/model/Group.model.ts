export interface ICheckboxOption {
  label: string;
  value: any;
  disabled?: boolean;
}

export interface ICheckboxGroupProps {
  options?: Array<ICheckboxOption>;
  value?: Array<any>,
  onChange?: (values: Array<any>) => void;
}