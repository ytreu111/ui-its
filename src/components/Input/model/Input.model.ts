import { ReactNode } from 'react';

export interface InputProps {
  value?: string;
  disabled?: boolean;
  small?: boolean;
  placeholder?: string;
  editableHelp?: string;
  label?: string;
  icon?: ReactNode;
  error?: boolean;
  autoWidth?: boolean;
  onChange?: (value: string) => void;
  onBlur?: () => void;
  onFocus?: () => void;
  name?: string;
  width?: number | string;
  optional?: () => void;
  optionalIcon?: ReactNode;
}
