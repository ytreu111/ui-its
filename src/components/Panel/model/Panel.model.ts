import { CSSProperties, ReactNode } from "react";

export interface IPanelProps{
  className?: string;
  style?: CSSProperties;
  headerContent?: ReactNode;
  headerStyle?: CSSProperties;
  headerClass?: string;
  headerComponent?: ReactNode;
  footerContent?: ReactNode;
  footerStyle?: CSSProperties;
  footerClass?: string;
  footerComponent?: ReactNode;
  contentStyle?: CSSProperties;
  contentClass?: string;
}