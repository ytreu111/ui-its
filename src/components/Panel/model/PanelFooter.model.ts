import { CSSProperties } from "react";

export interface IPanelFooterProps {
  className?: string;
  style?: CSSProperties;
}