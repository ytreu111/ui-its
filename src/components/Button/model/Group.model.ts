import { StyleType } from './Button.model';
import { ReactNode } from 'react';

export interface IButtonGroupProps {
  value?: string;
  buttonGap?: number;
  selectButton?: StyleType,
  deselectButton?: StyleType,
  defaultIcon?: ReactNode,
  onChange?: (event: MouseEvent, value?: string) => void;
  defaultValue?: string;
}

type ButtonOptionsFields = 'selectButton' | 'deselectButton' | 'defaultIcon' | 'onChange';

export type ButtonOptions = Required<Pick<IButtonGroupProps, ButtonOptionsFields>>;