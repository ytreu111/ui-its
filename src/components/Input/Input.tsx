import React, { FC, useCallback, useMemo, useState } from 'react';
import { Input as InputAntd } from 'antd';
import { ThemeProvider } from 'styled-components';

import { Styled } from './style/Input.styled';
import { InputProps } from './model/Input.model';
import { theme } from 'styles';

const {
  StyledInputContainer,
  StyledInputWrap,
  StyledInputEditablePlaceholder,
  StyledInputLabel,
} = Styled;

const Input: FC<InputProps> = (
  {
    value = '',
    disabled,
    small,
    placeholder,
    editableHelp = 'Введите данные',
    label,
    icon,
    error,
    allowClear,
    autoWidth,
    onChange,
    onBlur,
    onFocus,
    name,
  }) => {
  const [visiblePlaceholder, setVisiblePlaceholder] = useState(false);
  const [inputValue, setInputValue] = useState(value);

  const showClearButton = useMemo(
    () => !!(allowClear && !!inputValue && !disabled),
    [allowClear, disabled, inputValue],
  );

  const placeholderText = useMemo(
    () => (visiblePlaceholder ? editableHelp : placeholder)
    , [editableHelp, placeholder, visiblePlaceholder]);

  const classNameWrap = useMemo(() => {
    const className: Array<string> = [];

    if (small) className.push('small');

    if (autoWidth) className.push('auto');

    if (icon) className.push('icon');

    if (visiblePlaceholder) className.push('focus');

    if (visiblePlaceholder || !!inputValue) className.push('placeholder_top');

    if (error) className.push('error');

    if (showClearButton) className.push('clear');

    if (disabled) className.push('disabled');

    return className.join(' ');
  }, [autoWidth, disabled, error, icon, inputValue, showClearButton, small, visiblePlaceholder]);

  const onFocusInput = useCallback(() => {
    setVisiblePlaceholder(true);
    if (onFocus) onFocus();
  }, [onFocus]);

  const onBlurInput = useCallback(() => {
    setVisiblePlaceholder(false);
    if (onBlur) onBlur();
  }, [onBlur]);

  const onChangeInput = useCallback((e) => {
    setInputValue(e.currentTarget.value);
    if (onChange) onChange(e.currentTarget.value);
  }, [onChange]);

  return (
    <ThemeProvider theme={theme}>
      <StyledInputContainer>
        <StyledInputWrap className={classNameWrap}>
          {!small && <StyledInputEditablePlaceholder>{placeholder}</StyledInputEditablePlaceholder>}
          <InputAntd
            onFocus={onFocusInput}
            onBlur={onBlurInput}
            disabled={disabled}
            prefix={icon}
            value={inputValue}
            placeholder={placeholderText}
            onChange={onChangeInput}
            name={name}
          />
        </StyledInputWrap>
        {label && <StyledInputLabel>{label}</StyledInputLabel>}
      </StyledInputContainer>
    </ThemeProvider>
  );
};

export default Input;
