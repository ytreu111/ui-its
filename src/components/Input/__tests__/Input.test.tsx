import { render, screen, prettyDOM, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '../Input';
import { StarIcon } from 'ui-its-icons';
import { InputTestElemsEnum } from './Input.test.model';

const onChange = jest.fn();
const onFocus = jest.fn();
const onBlur = jest.fn();
const optional = jest.fn();

describe('Компонент Input', () => {
  test('className', () => {
    const { baseElement } = render(
      <Input
        className={'test'}
        small
        error
        icon={<StarIcon />}
      />,
    );
    expect(prettyDOM(baseElement)).toMatchSnapshot();
  });
  test('disabled', () => {
    render(<Input disabled />);
    const input = screen.getByTestId('input');
    userEvent.click(input);
    expect(input).not.toHaveFocus();
    expect(input).toBeDisabled();
  });
  test('focus/blur', () => {
    const { baseElement } = render(
      <Input
        onFocus={onFocus}
        onBlur={onBlur}
      />,
    );

    const input = screen.getByTestId(InputTestElemsEnum.input);
    const inputSnapshot = prettyDOM(baseElement);

    userEvent.click(input);
    expect(input).toHaveFocus();
    expect(onFocus).toHaveBeenCalledTimes(1);

    fireEvent.focusOut(input);
    expect(inputSnapshot).toBe(prettyDOM(baseElement));
    expect(onBlur).toHaveBeenCalledTimes(1);
  });
  test('Функционал кнопки(optional)', () => {
    const { rerender } = render(
      <Input
        optional={optional}
      />,
    );

    const optionalIcon = screen.getByTestId(InputTestElemsEnum.optional);
    expect(optionalIcon).toBeInTheDocument();
    userEvent.click(optionalIcon);
    expect(optional).toHaveBeenCalledTimes(1);

    rerender(<Input
      optionalIcon={<StarIcon data-testid={'optional-test'} />}
      optional={optional}
    />);
    const testOptionalIcon = screen.getByTestId('optional-test');
    expect(testOptionalIcon).toBeInTheDocument();
    userEvent.click(testOptionalIcon);
    expect(optional).toHaveBeenCalledTimes(2);
  });
  test('Тип', () => {
    const { rerender } = render(<Input />);
    expect(screen.getByTestId(InputTestElemsEnum.input)).toBeInTheDocument();
    rerender(<Input type={'password'} />);
    expect(screen.getByTestId(InputTestElemsEnum.inputPassword)).toBeInTheDocument();
  });
  test('label', () => {
    render(<Input label={'input'} />);
    expect(screen.getByText('input')).toBeInTheDocument();
  });
  test('onChange', () => {
    render(<Input onChange={onChange} />);
    userEvent.click(screen.getByTestId(InputTestElemsEnum.input));
    userEvent.keyboard('input');
    expect(onChange).toHaveBeenCalledTimes(5);
  });
});