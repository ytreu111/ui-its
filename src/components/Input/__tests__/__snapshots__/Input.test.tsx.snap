// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`–ö–æ–º–ø–æ–Ω–µ–Ω—Ç Input className 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-bdvvtL bcaXYY\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-eCImPb eFJDNR test small icon error\\"[39m
      [36m>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"ant-input-affix-wrapper\\"[39m
        [36m>[39m
          [36m<span[39m
            [33mclass[39m=[32m\\"ant-input-prefix\\"[39m
          [36m>[39m
            [36m<span[39m
              [33mclass[39m=[32m\\"anticon\\"[39m
              [33mrole[39m=[32m\\"img\\"[39m
            [36m>[39m
              [36m<svg[39m
                [33mfill[39m=[32m\\"currentColor\\"[39m
                [33mheight[39m=[32m\\"24\\"[39m
                [33mviewBox[39m=[32m\\"0 0 24 24\\"[39m
                [33mwidth[39m=[32m\\"24\\"[39m
                [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
              [36m>[39m
                [36m<path[39m
                  [33md[39m=[32m\\"M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z\\"[39m
                [36m/>[39m
              [36m</svg>[39m
            [36m</span>[39m
          [36m</span>[39m
          [36m<input[39m
            [33mclass[39m=[32m\\"ant-input\\"[39m
            [33mdata-testid[39m=[32m\\"input\\"[39m
            [33mtype[39m=[32m\\"text\\"[39m
            [33mvalue[39m=[32m\\"\\"[39m
          [36m/>[39m
        [36m</span>[39m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
